feat: Spatial Navigation

bzizmo opened the pull request:
RFC for Spatial Navigation

Commit:
Added proposal for Spatial navigation

Commit:
Fix headers

Commit:
Update filename and header with PR number

bzizmo committed:
Delete .idea directory

misteroneill commented on proposals/0002-spatial-navigation.md:
```diff
@@ -0,0 +1,143 @@
+[drawbacks]: #drawbacks
+
+- There might be some performance issues, especially with lower-end devices, due to the polyfill usage.
+- There might also be compatibility issues with certain platforms, especially older ones.
```
Our minimum browser engine support in Video.js 8 is Chromium 53. Of course, that doesn't help with older devices and their hardware limitations.

bzizmo replied:
Thanks for pointing this out. I'll remove these drawbacks as they are related to the spatial-navigation polyfill which will not be included.

bzizmo replied:
Changed in the latest commit

misteroneill commented on proposals/0002-spatial-navigation.md:
```diff
@@ -0,0 +1,143 @@
+
+This new feature will follow these steps:
+- Introduce a public option named `enableRCUNavigation`.
+- When a player is set up with the rcu navigation feature active, initiate the SpatialNavigation instance.
```
What is the workflow for users who may already be using Video.js in a Smart TV context?

I'm thinking here that they may have implemented their own spatial navigation handling in their web app. Would they simply leave the option disabled/inactive?

dzianis-dashkevich replied:
I believe they simply keep it disabled (simply do nothing, because it is disabled by default) if they use video.js only for playback and do not use video.js UI or if they've already implemented integration for their app's spatial navigation solution and video.js UI.

bzizmo replied:
This new feature will not impact in any way the current Video.js users if they keep it disabled.

misteroneill replied:
Yeah, that was my expectation, but it might be good to specifically mention it here.

bzizmo replied:
Added with the latest commit.

misteroneill commented on proposals/0002-spatial-navigation.md:
```diff
@@ -0,0 +1,143 @@
+[reference-level-explanation]: #reference-level-explanation
+
+## Dependency Management and Configuration Options
+For this feature, additional dependencies will be added: spatial-navigation-polyfill and flat-polyfill. Flat-polyfill is needed to accommodate devices lacking support for Array.prototype.flat.
```
In general, we have traditionally avoided bundling polyfills with Video.js. Instead, we've preferred to allow users to supply a polyfill of their choosing (if any). This was how we handled `Promise` in earlier versions of Video.js.

The reasoning behind this is that Video.js does not need to be in the business of changing native APIs implicitly.

I'd suggest that Video.js can feature detect the spatial navigation API(s) before trying to do anything with them. Methods that depend on native spatial navigation (or a polyfill) would simply become no-ops when that support does not exist.

gkatsev replied:
Another reason not to include the polyfill by default is filesize.
The polyfill is 72K ungzipped and 14K gzipped. (It could be a bit smaller if also minified). 14K isn't an insignificant amount to add.

(When minified, it goes down to 34K and 7K mingzipped)

----------------

> Methods that depend on native spatial navigation (or a polyfill) would simply become no-ops when that support does not exist.

👍🏻

dzianis-dashkevich replied:
> Methods that depend on native spatial navigation (or a polyfill) would simply become no-ops when that support does not exist.

As far as I understand, there is no native spatial navigation yet. There are no platforms (browsers) that officially implemented this spec: https://www.w3.org/TR/css-nav-1/

Please correct me if I am wrong!

This means that users use this polyfill, other libraries, or a custom implementation for spatial navigation.

Few thoughts:

**Option-1**
If we want to provide an out-of-the-box experience, we may consider this polyfill as a library, meaning we do not pollute global scope but rather import specific functions in our code base and use them.
Consider it as the following:

```js
import { navigate } from 'navigation-lib';

// use navigate in video.js. This pure function will be bundled with video.js
```

**Option-2**
Or we can provide a mechanism to set spatial navigation solutions. This spatial navigation solution should implement a specific interface that we expect. In this scenario, users will have to implement adapters around their spatial navigation solution to conform to the expected interface.


> Another reason not to include the polyfill by default is filesize.

We can provide an alternative bundle if spatial-navigation-related stuff is significant.

bzizmo replied:
Thanks for the feedback. I understand your concerns about adding an additional polyfill. I'll remove all references to it.
@dzianis-dashkevich that's correct. Currently, there is no native spatial navigation on Samsung Tizen and LG webOS platforms.

misteroneill replied:
I think it's right for Video.js to require developers include the polyfill separately.

I don't think we want to be in the business of providing multiple alternative builds. I don't think it adds enough benefit to our users to include the polyfill for them in Video.js.

My expectation would be that anyone hoping to use Video.js in a Tizen/webOS app would probably be using many other libraries as well and will likely need the spatial navigation polyfill (or their own implementation) baked into their app even where Video.js is not used. So, if it were included in Video.js in these cases, it would make their lives more difficult, not less.

We should, however, provide some kind of a sample implementation or guide document on videojs.com!

dzianis-dashkevich replied:
@misteroneill 
Does it mean you think that it will be better to go with something like option 2 from my message above?

misteroneill replied:
@dzianis-dashkevich Yes, I think something like option 2 would be best.

We can model the interface we expect on the standard interface (if such exists) similar to the way Video.js techs model their interface on the `HTMLMediaElement` interface.

misteroneill commented on proposals/0002-spatial-navigation.md:
```diff
@@ -0,0 +1,143 @@
+## Code and UI changes
+
+This new feature will follow these steps:
+- Introduce a public option named `enableRCUNavigation`.
```
Should we name this `enableSpatialNavigation` instead? I think the RCU abbreviation is fairly obscure.

bzizmo replied:
I'll change the name.

misteroneill commented on proposals/0002-spatial-navigation.md:
```diff
@@ -0,0 +1,143 @@
+
+### In terms of changes the UI
+- Since LEFT and RIGHT keys are reserved for video seeking, we’ll have to change the UI and move the progress bar on top of the buttons in order to allow the users to use those keys to navigate between elements when buttons are focused and seek the video when the progress bar is focused
+- Current player controls are too small for big screens, so we’re aiming to make those bigger
```
Agreed. We'll tackle that outside this feature, but good to call it out! 👍

bzizmo replied:
OK, since changing the UI is out of scope, I'd like to propose another approach for those who want to keep the same Video.js UI, and still want to use spatial navigation.

Introduce `spatialNavigation` option (instead of enableSpatialNavigation) as an object containing two properties:
- `enabled`, which can be set to `true` or` false to enable or disable the spatial navigation feature,
- seek, which can be set to either horizontal or vertical

- When seek is set to `horizontal`, only the LEFT and RIGHT arrow keys will invoke the stepForward and stepBack functions, permitting users to progress or regress the playback timeline. On the other hand, the UP and DOWN arrow keys will facilitate navigation out of the Slider, transitioning focus to other UI elements.
- If seek is defined as `vertical`, the roles of the arrow keys will be inverted. The UP and DOWN arrow keys will invoke the stepForward and stepBack functions, while the LEFT and RIGHT keys will be reserved for spatial navigation.

misteroneill replied:
Ok, cool. I like that idea. So, to make sure it's understood, the two options would be:
```
spatialNavigation: {
  enabled: true,
  seek: 'horizontal'
}
```
This would be used when you position the progress bar above the control bar - as in the Brightcove Player.
```
spatialNavigation: {
  enabled: true,
  seek: 'vertical'
}
```
This would be used when you position the progress bar in its default position within the control bar - as in Video.js.

Does this mean `vertical` would be the default?

Food for thought: if there are only two possible values for `seek`, should it instead be another boolean, maybe `horizontalSeek: [true|false]`? At a minimum, I'd suggest renaming it to `seekDirection: '[horizontal|vertical]'` to be a bit more semantically meaningful.

gkatsev replied:
This could be a good candidate for a follow-up RFC to fully think through all the potential issues with making the controls big for big screens.

bzizmo replied:
I like your idea @misteroneill to be boolean, makes more sense. 
Changed accordingly in the last commit

misteroneill commented on proposals/0002-spatial-navigation.md:
````diff
@@ -0,0 +1,143 @@
+}
+```
+
+### In terms of changes the UI
````
In order to ensure backward compatibility, we should only do this if the feature is enabled.

misteroneill reviewed (changes requested):
This is great, Borut, thank you for putting it together!

I think we should revisit the idea of including polyfill(s) _in Video.js_ - though, wherever we use this feature within Brightcove, we'd almost certainly want to include those polyfills.

Commit:
Removed the mention of the implementation of polyfills
Added a new proposal for handling different player layout navigation
Changed option 'enableRCUNavigation' to 'spatialNavigation' object with two properties (enabled and seek)
Removed changes to the UI
Changed drawbacks since the polyfills won't be included in the feature

Commit:
Update spatialNavigation property name to 'horizontalSeek'

Commit:
Merge remote-tracking branch 'origin/main'

misteroneill reviewed (approved):
I'm good with where this RFC is today.

The interface discussed in the comment thread above (potentially an interface between spatial navigation polyfill(s) and Video.js) could be elaborated on or left as an implementation detail. It doesn't seem functionally substantial enough to elaborate here.

dzianis-dashkevich commented:
Sound like we are good to merge it, Thank you @bzizmo
